ES|QL kql function.

afoucret opened the pull request:
This PR introduces support for the KQL query language as an ES|QL function.
Our main goal is to provides users a way forward to migrate from KQL to ES|QL by integrating KQL queries directly into ES|QL queries.


We recently added the support of KQL directly form the query DSL:

```
GET movies/_search
{
  "query": {
    "kql": {
      "query": "actors: \"Kurt Russel\" AND title: \"Big trouble\""
    }
  }
}
```

This PR expose this new capacity of the query DSL through a new `KQL` fulltext filtering function 

```
POST _query
{
  "query": """
    FROM movies
    | WHERE kql("actors: \"Kurt Russel\" AND title: \"Big trouble\"")
  """
}
```

The `KQL` function is conceptually very similar to the `QSTR` function and come with the same limitations.


The new function is enabled only in snapshots builds (same for the KQL query in the query DSL)

Commit:
Adding KqlQuery to esql-core query dsl supported queries.

Commit:
Adding KQL_FUNCTION capability.

Commit:
Change visibility for KqlQueryBuilder parser fields.

Commit:
Creating the KQL fulltext function.

Commit:
Register the KQL function in various places.

Commit:
Refactoring dependencies on KQL

afoucret committed:
Merge branch 'elastic:main' into esql-kql-function

Commit:
Remove moved tests.

Commit:
Adding some Verifier tests.

Commit:
Fixing telemetry.

Commit:
Add integration test for the KQL function.

Commit:
Adding CSV tests.

Commit:
Regenerate documentation for the KQL function.

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into esql-kql-function

Commit:
Add LogicalPhysicalPlanOptimizer test cases for KQL.

Commit:
Fixing some errors in the qstr csv test spec.

Commit:
Fix in the doc.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

Commit:
Lint fix.

afoucret commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:
```diff
@@ -101,8 +101,8 @@ book_no:keyword | title:text
 
-matchMultivaluedTextField
-required_capability: match_function
+qstrWithMultivaluedTextField
```
Not related to this PR but fixed few errors in the qstr csv spec

ioanatia commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java:
```diff
@@ -255,10 +255,10 @@ static boolean canPushToSource(Expression exp, LucenePushdownPredicates lucenePu
             return canPushSpatialFunctionToSource(spatial, lucenePushdownPredicates);
         } else if (exp instanceof StringQueryPredicate) {
             return true;
```
unrelated to your changes @afoucret , but @carlosdelest I wonder if this check is still needed? the `:` operator is using the match function, so I think we can remove?

afoucret replied:
I am not sure which condition became unnecessary. 

My understanding is that we can remove this one.

```
     if (exp instanceof StringQueryPredicate) {
          return true;
     } 
```

Can you confirm?

carlosdelest replied:
That's already removed in Ioana's PR - no need to do anything here 👍

ioanatia commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1261,11 +1261,68 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+            "1:77: [KQL] function cannot be used after RENAME",
+            error("from test | STATS c = COUNT(emp_no) BY languages | rename c as total_emps | where kql(\"Anna\")")
+        );
+        assertEquals(
+            "1:54: [KQL] function cannot be used after KEEP",
+            error("from test | rename last_name as name | keep emp_no | where kql(\"Anna\")")
+        );
```
this looks like a duplicate. we already test whether it can be used after `KEEP` in one of the assertions above. maybe we can check for `DROP` here instead?

afoucret replied:
Now using a drop.

Commit:
Small test change.

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into esql-kql-function

Commit:
Fix regression from merge.

Commit:
Fix regression from merge.

Commit:
Improving tests.

carlosdelest reviewed (approved):
Great job! 👏 

Some comments about potential refactors to unify code with Qstr - but not needed to be done as part of this PR.

Commit:
Ensure tests will not break on snapshot builds.

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into esql-kql-function

afoucret commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/plugin/KqlFunctionIT.java:
```diff
@@ -0,0 +1,146 @@
+
+    @Override
+    protected EsqlQueryResponse run(EsqlQueryRequest request) {
+        assumeTrue("kql function capability not available", EsqlCapabilities.Cap.KQL_FUNCTION.isEnabled());
```
Now in a before class.

Commit:
Refactor fulltext functions tests to avoid code duplication.

afoucret commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/KqlTests.java:
```diff
@@ -0,0 +1,74 @@
+import static org.hamcrest.Matchers.equalTo;
+
+@FunctionName("kql")
+public class KqlTests extends AbstractFunctionTestCase {
```
Done!

fang-xing-esql reviewed (commented):
Thanks for putting these together @afoucret ! It looks pretty good to me, I added some comments.

Here are my understanding of the `KQL` function, and please correct me if I missed anything:

- In terms of the interface, the `KQL` function is very similar to `QSTR` so it makes sense to follow a similar approach as `QSTR` to support it in the `ES|QL`. 
- However they are different in syntax according to this [google doc](https://docs.google.com/document/d/1bHh2koNFCtymT91y4YVdQQisdRY6-4IvY14sIpDGzbc/edit?tab=t.0#heading=h.1vk1ttronc2o), although their syntax look very similar to me, I won't be surprised that I might make mistakes on the syntax when using them.  From users perspective, it will be helpful to have a page(or a link to existing ES docs) in the `ES|QL` doc to describe when to use `KQL` and when to use `QSTR`, and have examples to describe the usages. Especially they are now both available through `ES|QL` we may want to suggest the best practice to reduce confusions. Perhaps it is only me, but I'm not clear when I should use which function. This is not a needed for a snapshot function, but it will be helpful to get us prepared for a tech preview or GA. 
- I suggest to add ES|QL-ui label to notify kibana about this new function, it is near the FF of 8.17, it will be nice to give them a heads up if this is targeting to 8.17.
- I suggest to add test-release label to run this through release tests, sometimes it does catch surprise for us.

Commit:
Remove duplicated Verifier code for QSTR and KQL.

Commit:
Sort functions by alphabetical order.

Commit:
Remove useless override.

Commit:
Remove useless annotation.

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into esql-kql-function

afoucret commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Kql.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public String functionName() {
+        return "KQL";
+    }
```
Removed the override.

afoucret commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:
```diff
@@ -409,7 +410,8 @@ private static FunctionDefinition[][] snapshotFunctions() {
                 def(Categorize.class, Categorize::new, "categorize"),
-                def(Rate.class, Rate::withUnresolvedTimestamp, "rate") } };
+                def(Rate.class, Rate::withUnresolvedTimestamp, "rate"),
+                def(Kql.class, Kql::new, "kql") } };
```
Done. Thank you.

elasticsearchmachine commented:
Pinging @elastic/kibana-esql (ES|QL-ui)

afoucret commented:
@fang-xing-esql Thanks for the detailed review and feedback!

You're absolutely right that the `KQL` and `QSTR` functions are quite similar.
Both allow to leverage the `_search` infrastructure to execute another query language (respectively the Kibana Query Language and Lucene query string). 

That said, for KQL, our intention is not to promote it as a standalone feature but rather to facilitate users migrating to ES|QL. This is why we’ve kept communication about it relatively low-key. 
I will work on enriching the documentation by adding a link to the existing KQL page.

Few other points:
- I updated the code per your suggestions.
- I added the `test-release` label to ensure those tests are sucessful
- I updated the version label to target `8.18.0` instead of `8.17.0`
- I added the `ES|QL-ui` label to keep the Kibana team informed of the progress

fang-xing-esql reviewed (approved):
Thank you @afoucret! LGTM. And thanks to @carlosdelest for reviewing as well.

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into esql-kql-function

Commit:
Remove gen files added by mistake

Commit:
Fix test compile after merge from main.

Commit:
Lint fix.

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into esql-kql-function

Commit:
Update the number of functions in telemetry.
